### PR TITLE
#3035 Convenience methods for sending .started/.status-changed/.finished events

### PR DESCRIPTION
**Pull request by bacherfl**
Closes https://github.com/keptn/keptn/issues/3035

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -8,6 +8,7 @@ require (
 	github.com/go-openapi/strfmt v0.19.3
 	github.com/go-openapi/swag v0.19.5
 	github.com/go-openapi/validate v0.19.4
-	github.com/stretchr/testify v1.7.0
-	gopkg.in/yaml.v2 v2.4.0
+	github.com/google/uuid v1.1.1
+	github.com/stretchr/testify v1.5.1
+	gopkg.in/yaml.v2 v2.2.8
```
Comment (christian-kreuzberger-dtx) [inline go.mod:13]: Please revert the changes, we upgraded testify and yaml using dependabot. And I'm not sure about why uuid was removed, but in case we do need it, please add it again.

File: pkg/lib/v0_2_0/keptn.go [hunk 2]
```diff
@@ -83,3 +87,106 @@ func (k *Keptn) GetShipyard() (*Shipyard, error) {
...
+
+// SendCloudEvent sends a cloudevent to the event broker
+func (k *Keptn) SendCloudEvent(event cloudevents.Event) error {
+	if k.UseLocalFileSystem {
+		log.Println(fmt.Printf("%v", string(event.Data())))
+		return nil
+	}
+
+	return k.EventSender.SendEvent(event)
+}
```
Comment (christian-kreuzberger-dtx) [inline pkg/lib/v0_2_0/keptn.go:99]: Isn't this a duplicate of https://github.com/keptn/go-utils/blob/67ac0f2ba06dd11f3e045f7ff7053674175fbbce/pkg/lib/v0_2_0/events.go#L123-L131

Comment (christian-kreuzberger-dtx) [reply pkg/lib/v0_2_0/keptn.go:99]: And it's not used in this PR either as far as I can see.

File: pkg/lib/v0_2_0/keptn.go [hunk 3]
```diff
@@ -83,3 +87,106 @@ func (k *Keptn) GetShipyard() (*Shipyard, error) {
...
+
+// SendTaskStartedEvent sends a .started event for the incoming .triggered event the KeptnHandler was initialized with.
+// It returns the ID of the sent CloudEvent or an error
+func (k *Keptn) SendTaskStartedEvent(data keptn.EventProperties, source string) (string, error) {
+
```
Comment (christian-kreuzberger-dtx) [inline pkg/lib/v0_2_0/keptn.go:104]: should we introduce a check here whether `k.CloudEvent != nil` to avoid wrong usage of this function? 

Same for `SendTaskStatusChangedEvent` and `SendTaskFinishedEvent`

File: pkg/lib/v0_2_0/keptn.go [hunk 4]
```diff
@@ -83,3 +87,106 @@ func (k *Keptn) GetShipyard() (*Shipyard, error) {
...
+	return k.sendEventWithBaseEventContext(data, source, err, outEventType)
+}
+
+// SendTaskStartedEvent sends a .status.changed event for the incoming .triggered event the KeptnHandler was initialized with.
+// It returns the ID of the sent CloudEvent or an error
+func (k *Keptn) SendTaskStatusChangedEvent(data keptn.EventProperties, source string) (string, error) {
+	outEventType, err := GetEventTypeForTriggeredEvent(k.CloudEvent.Type(), keptnStatusChangedEventSuffix)
+	if err != nil {
+		return "", fmt.Errorf("could not determine .started event type for base event: %s", err.Error())
```
Comment (christian-kreuzberger-dtx) [inline pkg/lib/v0_2_0/keptn.go:118]: copy paste error in error message, should be `.status.changed`

File: pkg/lib/v0_2_0/keptn.go [hunk 5]
```diff
@@ -83,3 +87,106 @@ func (k *Keptn) GetShipyard() (*Shipyard, error) {
...
+		return "", fmt.Errorf("could not initialize CloudEvent: %s", err.Error())
+	}
+
+	if err := k.EventSender.SendEvent(*ce); err != nil {
+		return "", fmt.Errorf("could not send CloudEvent: %s", err.Error())
+	}
+	return ce.ID(), nil
+}
+
+func (k *Keptn) createCloudEventWithContextAndPayload(outEventType string, keptnContext interface{}, source string, data keptn.EventProperties) (*cloudevents.Event, error) {
```
Comment (christian-kreuzberger-dtx) [inline pkg/lib/v0_2_0/keptn.go:155]: Please describe in a comment what this helper function is doing

Please also add inline comments, e.g., 
```
// use existing  sh keptn context for the new cloudevent
ce.SetExtension(keptnContextCEExtension, keptnContext)
```

File: pkg/lib/v0_2_0/keptn.go (hunk 2 above)
Comment (bacherfl) [reply pkg/lib/v0_2_0/keptn.go:99]: this function was moved because I figured it would make sense to have the SendCloudEvent method in the same place where the `SendStarted/StatusChanged/Finished` methods are

File: pkg/lib/v0_2_0/keptn.go (hunk 3 above)
Comment (bacherfl) [reply pkg/lib/v0_2_0/keptn.go:104]: makes sense, I will add a check for that

File: pkg/lib/v0_2_0/keptn.go [hunk 6]
```diff
@@ -83,3 +87,116 @@ func (k *Keptn) GetShipyard() (*Shipyard, error) {
...
+
+// SendTaskStartedEvent sends a .finished event for the incoming .triggered event the KeptnHandler was initialized with.
+// It returns the ID of the sent CloudEvent or an error
+func (k *Keptn) SendTaskFinishedEvent(data keptn.EventProperties, source string) (string, error) {
+	if k.CloudEvent == nil {
+		return "", fmt.Errorf("no incoming .triggered CloudEvent provided to the Keptn Handler")
+	}
+	outEventType, err := GetEventTypeForTriggeredEvent(k.CloudEvent.Type(), keptnFinishedEventSuffix)
+	if err != nil {
+		return "", fmt.Errorf("could not determine .started event type for base event: %s", err.Error())
```
Comment (christian-kreuzberger-dtx) [inline pkg/lib/v0_2_0/keptn.go:137]: shouldn't this be `.finished`?

File: go.mod (hunk 1 above)
Comment (bacherfl) [reply go.mod]: done, the dependencies are now up to date again

File: pkg/lib/v0_2_0/keptn.go (hunk 6 above)
Comment (bacherfl) [reply pkg/lib/v0_2_0/keptn.go]: good catch, is fixed now

**Review by christian-kreuzberger-dtx — APPROVED** [review]
LGTM - I'll test those changes soon :)
